packages/tailwindcss/src/functions.ts

```diff
@@ -0,0 +1,166 @@
+        )
+      }
+      let path = pathNode.value
+
+      // For the theme function arguments, we require all separators to contain
+      // comma (`,`), spaces alone should be merged into the previous word to
+      // avoid splitting in this case:
+      //
+      // theme(colors.red.500 / 75%) theme(colors.red.500 / 75%, foo, bar)
+      //
+      // We only need to do this for the first node, as the fallback values are
+      // passed through as-is.
+      let skipUntilIndex = 1
+      for (let i = skipUntilIndex; i < node.nodes.length; i++) {
+        if (node.nodes[i].value.includes(',')) {
+          break
+        }
+        path += node.nodes[i].value
+        skipUntilIndex = i
+      }
+
+      path = eventuallyUnquote(path)
+      let fallbackValues = node.nodes.slice(skipUntilIndex + 2)
+
+      replaceWith(cssThemeFn(pluginApi, path, fallbackValues))
+    }
+  })
+
+  return toValueCss(ast)
+}
+
+function cssThemeFn(
+  pluginApi: PluginAPI,
+  path: string,
+  fallbackValues: ValueAstNode[],
+): ValueAstNode[] {
+  let modifier: string | null = null
+  // Extract an eventual modifier from the path. e.g.:
+  // - "colors.red.500 / 50%" -> "50%"
+  // - "foo/bar/baz/50%"      -> "50%"
+  let lastSlash = path.lastIndexOf('/')
+  if (lastSlash !== -1) {
+    modifier = path.slice(lastSlash + 1).trim()
+    path = path.slice(0, lastSlash).trim()
+  }
+
+  let resolvedValue: string | null = null
+  let themeValue = pluginApi.theme(path)
+
+  if (Array.isArray(themeValue)) {
+    // When a tuple is returned, return the first element
+    resolvedValue = themeValue[0]
+    // We otherwise only ignore string values here, objects (and namespace maps)
+    // are treated as non-resolved values for the CSS `theme()` function.
+  } else if (typeof themeValue === 'string') {
+    resolvedValue = themeValue
+  }
+
+  // The plugin `theme()` function currently returns resolved values (so values
+  // that are wrapped in `var()` and the CSS variable name). This `theme()`
+  // function should, however, instead return the raw values. Raw values are
+  // necessary because they might be used in positions where `var()` is not
+  // supported like `@media (min-width: theme(--breakpoint-sm))`.
+  //
+  // Since the plugin `theme()` function operates on a materialized config
+  // object provided by plugins, the values would already be read from the CSS
+  // theme before we even get here (the config object is initialized together
+  // with the plugins). Subsequently, we can only read resolved values here
+  // unless we create a separate config object containing unresolved CSS values
+  // that is only for use with the CSS `theme()` function.
+  //
+  // Since this is overkill, we instead introspect the string and try to unwrap
+  // the `var()` call for now. This works because we always define a fallback
+  // value that points to the raw CSS variable value.
```

Comment:
Does this fail if someone registers something using `@theme inline` that includes a variable with a fallback?

```css
@theme inline {
  --color-primary: var(--brand-primary, #ff0);
}
```

In this case `theme(--color-primary)` should resolve `var(--brand-primary, #ff0)`, not just `#ff0`.

Reply:
Yep, it does. It's something that I also thought of today. We can probably be stricter and check for only `var(` calls with the CSS variable name that matches the theme key, but all of this still feels like we shouldn't do it to begin with. The alternative I’m thinking of is to use an unmaterialized view instead (so the `theme()` calls return objects and don't resolve to the theme values) but that also makes it needlessly complex.